File: duplex/cpp/src/pipe.cpp
#include <nightseam/duplex/conn.hpp>

#include <array>
#include <condition_variable>
#include <deque>
#include <mutex>

namespace nightseam::duplex {
namespace {

struct End {
    std::deque<Frame> sent;
    bool closed = false;
    int code = 1006;
    std::string reason;
};

struct Shared {
    std::mutex mutex;
    std::condition_variable_any changed;
    std::array<End, 2> ends;
    std::size_t byte_limit, capacity;
    Shared(std::size_t byte_limit, std::size_t capacity) : byte_limit(byte_limit), capacity(capacity) {}
};

template<class Predicate>
void await(Shared& state, std::unique_lock<std::mutex>& lock, Wait wait, Predicate ready) {
    wait.check();
    bool reached;
    if (wait.deadline == std::chrono::steady_clock::time_point::max()) {
        reached = state.changed.wait(lock, wait.stop, ready);
    } else {
        reached = state.changed.wait_until(lock, wait.stop, wait.deadline, ready);
    }
    if (!reached) { wait.check(); throw DeadlineExceeded(); }
}

class PipeEnd final : public Conn {
    std::shared_ptr<Shared> state_;
    std::size_t index_;

    void finish(int code, std::string reason) noexcept {
        {
            std::lock_guard lock(state_->mutex);
            auto& local = state_->ends[index_];
            if (local.closed) return;
            local.closed = true;
            local.code = code;
            local.reason = std::move(reason);
        }
        state_->changed.notify_all();
    }

public:
    PipeEnd(std::shared_ptr<Shared> state, std::size_t index) : state_(std::move(state)), index_(index) {}
    ~PipeEnd() override { abort(); }

    void send(const Frame& frame, Wait wait) override {
        if (frame.kind != Kind::text && frame.kind != Kind::binary) throw std::invalid_argument("duplex frame of no kind");
        std::unique_lock lock(state_->mutex);
        auto& local = state_->ends[index_];
        auto& remote = state_->ends[1-index_];
        if (local.closed) throw Closed();
        if (remote.closed) throw CloseError(remote.code, remote.reason);
        await(*state_, lock, wait, [&] { return local.closed || remote.closed || local.sent.size() < state_->capacity; });
        if (local.closed) throw Closed();
        if (remote.closed) throw CloseError(remote.code, remote.reason);
        local.sent.push_back(frame);
        lock.unlock();
        state_->changed.notify_all();
    }

    Frame receive(Wait wait) override {
        std::unique_lock lock(state_->mutex);
        auto& local = state_->ends[index_];
        auto& remote = state_->ends[1-index_];
        if (local.closed) throw Closed();
        await(*state_, lock, wait, [&] { return local.closed || remote.closed || !remote.sent.empty(); });
        if (local.closed) throw Closed();
        if (remote.sent.empty()) throw CloseError(remote.code, remote.reason);
        auto frame = std::move(remote.sent.front());
        remote.sent.pop_front();
        auto oversized = state_->byte_limit != 0 && frame.data.size() > state_->byte_limit;
        if (oversized) local.closed = true;
        lock.unlock();
        state_->changed.notify_all();
        if (oversized) throw FrameTooLarge("duplex frame exceeds the receive byte limit");
        return frame;
    }

    void close(int code, std::string reason, Wait) override { finish(code, std::move(reason)); }
    void abort() noexcept override { finish(1006, ""); }
};

} // namespace

std::pair<std::shared_ptr<Conn>, std::shared_ptr<Conn>> pipe(std::size_t byte_limit, std::size_t capacity) {
    if (capacity == 0) throw std::invalid_argument("a pipe's frame capacity must be positive");
    auto state = std::make_shared<Shared>(byte_limit, capacity);
    return {std::make_shared<PipeEnd>(state, 0), std::make_shared<PipeEnd>(state, 1)};
}

} // namespace nightseam::duplex

File: duplex/go/pipe.go
package duplex

import (
	"context"
	"fmt"
	bitwire "github.com/Bitspark/bitwire/wire/go"
	"sync"
)

// Pipe returns two connected ends in memory: what one sends, the other
// receives, in order. Each direction holds at most a few frames in flight,
// as a socket holds some bytes; past that a Send waits for a Receive, so
// the pipe carries backpressure as a transport does. A Close on one end is
// a CloseError on the other; an Abort is CodeAbnormalClosure there, as a
// dropped socket would be. It carries the duplex profile in tests without
// a socket, and it is the transport a protocol is held to before a real one
// is.
func Pipe(limit int64) (Conn, Conn) {
	a := &pipeEnd{limit: limit, frames: make(chan Frame, inFlight), done: make(chan struct{})}
	b := &pipeEnd{limit: limit, frames: make(chan Frame, inFlight), done: make(chan struct{})}
	a.remote, b.remote = b, a
	return a, b
}

// inFlight is how many frames a pipe holds per direction before a Send
// waits.
const inFlight = 8

type pipeEnd struct {
	limit  int64
	frames chan Frame // frames this end sends; the remote end receives from it
	remote *pipeEnd
	done   chan struct{}
	once   sync.Once
	mu     sync.Mutex
	closed *CloseError // what the remote will be told; nil after Abort
}

func (e *pipeEnd) end(closed *CloseError) {
	e.once.Do(func() {
		e.mu.Lock()
		e.closed = closed
		e.mu.Unlock()
		close(e.done)
	})
}

// remoteError is what this end returns once the remote end ended: its
// close, or an abnormal closure if it aborted.
func (e *pipeEnd) remoteError() error {
	e.remote.mu.Lock()
	defer e.remote.mu.Unlock()
	if e.remote.closed != nil {
		return &CloseError{Code: e.remote.closed.Code, Reason: e.remote.closed.Reason}
	}
	return &CloseError{Code: CodeAbnormalClosure}
}

func (e *pipeEnd) Send(ctx context.Context, frame Frame) error {
	if frame.Kind != Text && frame.Kind != Binary {
		return ErrNoKind
	}
	select {
	case <-e.done:
		return ErrClosed
	case <-e.remote.done:
		return e.remoteError()
	default:
	}
	data := make([]byte, len(frame.Data))
	copy(data, frame.Data)
	select {
	case e.frames <- Frame{Kind: frame.Kind, Data: data}:
		return nil
	case <-ctx.Done():
		return ctx.Err()
	case <-e.done:
		return ErrClosed
	case <-e.remote.done:
		return e.remoteError()
	}
}

func (e *pipeEnd) Receive(ctx context.Context) (Frame, error) {
	select {
	case <-e.done:
		return Frame{}, ErrClosed
	default:
	}
	deliver := func(frame Frame) (Frame, error) {
		if e.limit > 0 && int64(len(frame.Data)) > e.limit {
			_ = e.Abort()
			return Frame{}, fmt.Errorf("duplex frame of %d bytes exceeds the receive limit of %d", len(frame.Data), e.limit)
		}
		return frame, nil
	}
	select {
	case frame := <-e.remote.frames:
		return deliver(frame)
	case <-ctx.Done():
		return Frame{}, ctx.Err()
	case <-e.done:
		return Frame{}, ErrClosed
	case <-e.remote.done:
		// What the remote sent before it closed is still delivered, in order,
		// before its close is.
		select {
		case frame := <-e.remote.frames:
			return deliver(frame)
		default:
			return Frame{}, e.remoteError()
		}
	}
}

func (e *pipeEnd) Close(ctx context.Context, code bitwire.Code, reason string) error {
	e.end(&CloseError{Code: code, Reason: reason})
	return nil
}

func (e *pipeEnd) Abort() error {
	e.end(nil)
	return nil
}
